### PR TITLE
Add co soul contract

### PR DESCRIPTION
**Pull request by crabsinger**

### CODE REVIEW

File: contracts/cosoul/SBT.sol [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	function setSigner(address _signer) external onlyOwner {
+		signer = _signer;
+	}
+
+
+	function setCallers(address _caller, bool _val) external onlyOwner {
+		authorisedCallers[_caller] = _val;
+	}
+
+	function getSlot(uint8 _slot, uint256 _tokenId) public view returns(uint256 value) {
```
Comment (crabsinger) [inline contracts/cosoul/SBT.sol:45]: @OwlOfMoistness do you have some docs on how these work? I'd like to understand this a bit better.

File: contracts/cosoul/SBT.sol [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+		require(balanceOf(msg.sender) == 0);
+		require(mintNonces[msg.sender]++ == _nonce);
+		require(keccak256(abi.encodePacked(msg.sender, _nonce)).toEthSignedMessageHash().recover(_signature) == signer, "Sig not valid");
+		
+		_mint(msg.sender, ++counter);
+	}
+
+	/**
+	 * @notice
+	 * Function to mint token via signature to msg.sender
```
Comment (levity) [inline contracts/cosoul/SBT.sol:191]:
```suggestion
	 * Function to burn token via signature to msg.sender
```

File: contracts/cosoul/SBT.sol [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+		require(_slot < 8);
+		uint256 value = getSlot(_slot, _tokenId);
+		require(value >= _amount, "CoSoul: uint32 overflow");
+		uint256 current = blobs[_tokenId];
+		blobs[_tokenId] = current - (_amount << _slot);
+	}
+
+	/**
+	 * @notice
+	 * Function to sync blob data of a token from a merkle tree signed by our signer
```
Comment (crabsinger) [inline contracts/cosoul/SBT.sol:128]: is this using a merkle tree? can you explain that?

File: contracts/cosoul/SBT.sol [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		_transfer(_from, _to, _tokenId);
+	}
+
+	/**
+	 * @notice
+	 * Function to mint token via signature to msg.sender
+	 * @param _nonce Mint counter used to prevent replays
+	 * @param _signature Signature provided by our signer to validate mint
+	 */
+	function mintWithSignature(uint256 _nonce, bytes calldata _signature) external {
```
Comment (crabsinger) [inline contracts/cosoul/SBT.sol:181]: Can you explain this mintWithSignature, vs just a mint function that doesn't require a signature?

Comment (crabsinger) [reply contracts/cosoul/SBT.sol:181]: I'm assuming this pattern: https://portal.thirdweb.com/glossary/signature-based-minting

File: contracts/cosoul/SBT.sol (hunk 3 above)
Comment (crabsinger) [reply contracts/cosoul/SBT.sol:125]: no merkle tree. removed

File: contracts/cosoul/SBT.sol [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+	/**
+	 * @notice
+	 * Function to sync blob data of a token from a merkle tree signed by our signer
+	 * @param _data Blob data that will overwrite current data
+	 * @param _tokenId Token ID from which to update blob
+	 * @param _nonce Sync counter used to prevent replays
+	 * @param _signature Signature provided by our signer to validate leaf data
+	 */
+	function syncWithSignature(uint256 _data ,uint256 _tokenId, uint256 _nonce, bytes calldata _signature) external {
```
Comment (crabsinger) [inline contracts/cosoul/SBT.sol:131]: this is for user to sync data.

**Review by CryptoGraffe — APPROVED** [review]

**Pull request merged**